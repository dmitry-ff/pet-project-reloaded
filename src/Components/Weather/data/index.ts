import { MONTH_S } from "./MONTH_S";
import { WEEK_S } from "./WEEK_S";

export { MONTH_S, WEEK_S }