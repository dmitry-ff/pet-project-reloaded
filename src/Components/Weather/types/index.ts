import { TData } from "./TData";
import { THead } from "./THead";
import { TLoading } from "./TLoading";
import { TResponse } from "./TResponse";
import { TWeatherGetData } from "./TWeatherGetData";
import { TCanvas } from "./TCanvas";
import { TDataTemperature } from "./TDataTemperature";


export type {
  TData,
  THead,
  TLoading,
  TResponse,
  TWeatherGetData,
  TCanvas,
  TDataTemperature
}