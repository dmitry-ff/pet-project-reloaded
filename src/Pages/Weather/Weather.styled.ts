import styled from "@emotion/styled";

export const Heading = styled.h2`
    margin-top:50px;
    font-size: 2.5em;
    text-align: center;
    color: rgb(66, 66, 66);
`