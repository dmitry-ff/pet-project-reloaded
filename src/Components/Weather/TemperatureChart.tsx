import _ from "lodash";
import React from "react";
import { MONTH_S } from "../../Data/Consts";
import { Line } from 'react-chartjs-2';
import { TLoading } from '../../Data/Types';
import { EmotionJSX } from "@emotion/react/types/jsx-namespace";
type TData = {
  data: {
    date: string,
    day: {
      maxtemp_c: number,
      mintemp_c: number,
      avgtemp_c: number,
      maxwind_kph: number,
      daily_chance_of_rain: number,
      daily_chance_of_snow: number,
      condition: {
        text: string,
        icon: string
      }
    }
  }[]
}
const TemperatureChart: React.FC<TLoading & TData> = ({ loading, data }): EmotionJSX.Element => {

  const [chartData, setData] = React.useState({
    labels: _.map(data, item => {
      return `${MONTH_S[new Date(item.date).getMonth()]}, ${new Date(item.date).getDate()}`
    }),
    datasets: [
      {
        label: 'max t, °C',
        data: _.map(data, item => {
          return item.day.maxtemp_c;
        }),
        backgroundColor: 'rgba(222, 99, 132, 0.2)',
        borderColor: 'rgba(255, 99, 132, 1)',
        borderWidth: 2,
        tension: 0.3
      },
      {
        label: 'avg t, °C',
        data: _.map(data, item => {
          return (item.day.maxtemp_c + item.day.mintemp_c) / 2;
        }),
        backgroundColor: 'rgba(48, 207, 113,0.2)',
        borderColor: 'rgba(48, 207, 113,1)',
        borderWidth: 2,
        tension: 0.3

      },
      {
        label: 'min t, °C',
        data: _.map(data, item => {
          return item.day.mintemp_c;
        }),
        backgroundColor: 'rgba(1, 99, 132, 0.2)',
        borderColor: 'rgba(1, 99, 132, 1)',
        borderWidth: 2,
        tension: 0.3
      },
    ],
  });

  return <>{!loading && <Line data={chartData} />
  }</>
}

export default TemperatureChart;
// export { };