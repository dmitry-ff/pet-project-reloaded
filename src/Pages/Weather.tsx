import React from 'react';
import WeatherData from '../Components/Weather/WeatherGetData'
import { Button, Box, Link, TextField } from '@mui/material';
import styled from '@emotion/styled';
import { EmotionJSX } from '@emotion/react/types/jsx-namespace';
import { useFormik } from "formik";
import { Context } from '../Context';
import { TContext } from '../Data/Types';

const Heading = styled.h2`
    margin-top:50px;
    font-size: 2.5em;
    text-align: center;
    color: rgb(66, 66, 66);
`
interface IFormik {
  searchField: string,
}


const Weather: React.FC = (): EmotionJSX.Element => {

  const [url, setUrl] = React.useState<string>('');
  const [press, setPress] = React.useState<boolean>(false);
  const [town, setTown] = React.useState<string>('')
  const { context, setContext } = React.useContext(Context) as TContext;

  const validate = (values: IFormik,) => {
    let errors = {} as IFormik;
    if (!values.searchField) {
      errors.searchField = 'Required'
    } else if
      (!/^([a-zа-яё]+)$/i.test(values.searchField)) {
      errors.searchField = 'Incorrect Entry'
    }
    return errors;
  };

  const formik = useFormik({
    initialValues: {
      searchField: '',
    },
    validate,

    onSubmit: (values, { resetForm }) => {
      setUrl(url => url =
        `http://api.weatherapi.com/v1/forecast.json?key=3e2b23202a804627a5c115738222701&q=${formik.values.searchField}&days=3&aqi=no&alerts=no`);
      setPress(true);
      setTown(values.searchField.slice(0, 1).toUpperCase() + values.searchField.slice(1));
      resetForm();
      setContext(true)
    }
  })

  return (
    <Box sx={{
      display: 'flex',
      flexDirection: 'column',
      alignItems: 'center',
      width: "100%",
    }} >
      <Heading>Where do you whant to check weather?</Heading>
      <form onSubmit={formik.handleSubmit}>
        <Box sx={{
          width: '40em',
          height: 150,
          mt: 3,
          display: 'flex',
          flexDirection: 'column',
          justifyContent: 'space-evenly'
        }}>
          <TextField
            placeholder='...maybe in Paris?'
            autoComplete='off'
            id='searchField'
            name='searchField'
            value={formik.values.searchField}
            onChange={formik.handleChange}
            error={(formik.touched.searchField && Boolean(formik.errors.searchField)) || (!context)}
            helperText={(formik.touched.searchField && formik.errors.searchField) || (!context && 'Incorrect Entry')}

          />

          <Button
            type='submit'
            sx={{
              width: 'fit-content',
              alignSelf: 'center',
            }}
            color='primary'
            variant="contained">
            Check weather
          </Button>
        </Box>
      </form>
      {press && <WeatherData apiURL={url} townName={town} />}
      <span>
        {'Powered by '}
        <Link
          href="https://www.weatherapi.com/">WeatherAPI.com
        </Link>
      </span>
    </Box>
  );
}

export default Weather;
