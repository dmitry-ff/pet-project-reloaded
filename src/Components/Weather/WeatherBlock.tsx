import _, { isEmpty } from 'lodash';
import React from 'react';
import { WEEK_S } from "../../Data/Consts/WEEK_S";
import { WeatherData, WeatherToday, FirstColumn, SecondColumn, FourthColumn, WeatherWeek, WeatherCard } from '../../Style/Weather'
import { TLoading } from "../../Data/Types/TLoading";
import { EmotionJSX } from '@emotion/react/types/jsx-namespace';
type THead = {
  nameHead: string;
}
type TData = {
  data: {
    date: string,
    day: {
      maxtemp_c: number,
      mintemp_c: number,
      avgtemp_c: number,
      maxwind_kph: number,
      daily_chance_of_rain: number,
      daily_chance_of_snow: number,
      condition: {
        text: string,
        icon: string
      }
    }
  }[]
}

const WeatherBlock: React.FC<TLoading & THead & TData> = ({ loading, nameHead, data }): EmotionJSX.Element | null => {
  console.log(loading, 'LOADING')
  console.log(data, 'DATA');

  if (!loading && isEmpty(data)) return null

  return (
    <WeatherData>
      <span >
        Weather in <b>{nameHead}</b>
      </span>
      <WeatherToday>
        <FirstColumn>
          <span>{_.head(data)!.day.avgtemp_c > 0 ? '+' + _.floor(_.head(data)!.day.avgtemp_c) : _.floor(_.head(data)!.day.avgtemp_c)}°</span>
          <img src={_.head(data)!.day.condition.icon} alt={_.head(data)!.day.condition.text} />
        </FirstColumn>
        <SecondColumn>
          <span>{_.head(data)!.day.condition.text}</span>
          <div>
            <span>max <b>{_.head(data)!.day.maxtemp_c > 0 ? '+' + _.floor(_.head(data)!.day.maxtemp_c) : _.floor(_.head(data)!.day.maxtemp_c)}</b></span>
            <span>min <b>{_.head(data)!.day.mintemp_c > 0 ? '+' + _.floor(_.head(data)!.day.mintemp_c) : _.floor(_.head(data)!.day.mintemp_c)}</b></span>
          </div>
        </SecondColumn>
        <div>
          <span>Wind speed  <b>{(_.head(data)!.day.maxwind_kph / 3.6).toFixed(1)} m/s</b></span>
        </div>
        <FourthColumn>
          <span>{_.head(data)!.day.daily_chance_of_rain} %</span>
          <span>{_.head(data)!.day.daily_chance_of_snow} %</span>
        </FourthColumn>
      </WeatherToday>

      <WeatherWeek>
        {!loading && <>
          {
            _.map(data, (item, index) =>
              <div key={index}>
                <div>
                  {index === 0 ?
                    'Today'
                    : `${WEEK_S[new Date(item.date).getDay()]}, ${new Date(item.date).getDate()} `}
                </div>
                <WeatherCard>
                  <div>
                    <span>{item.day.maxtemp_c > 0 ? '+' + _.floor(item.day.maxtemp_c) : _.floor(item.day.maxtemp_c)}°</span>
                    <span>{item.day.mintemp_c > 0 ? '+' + _.floor(item.day.mintemp_c) : _.floor(item.day.mintemp_c)}°</span>
                  </div>
                  <img src={item.day.condition.icon} alt={item.day.condition.text} />
                </WeatherCard>
              </div>)
          }
        </>
        }

      </WeatherWeek>
    </WeatherData>
  )

}

export default WeatherBlock;
// export { }