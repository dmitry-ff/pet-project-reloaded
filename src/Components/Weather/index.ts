import TemperatureChart from "./TemperatureChart/TemperatureChart";
import VanillasCanvasChart from "./VanillasCanvasChart/VanillasCanvasChart";
import WeatherBlock from "./WeatherBlock/WeatherBlock";
import WeatherFetchData from "./WeatherFetchData/WeatherFetchData";

export {
  TemperatureChart,
  VanillasCanvasChart,
  WeatherBlock,
  WeatherFetchData,
}