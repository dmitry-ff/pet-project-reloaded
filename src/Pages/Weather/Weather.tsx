import React from 'react';
import { WeatherFetchData } from '../../Components/'
import { Button, Box, Link, TextField } from '@mui/material';
import { EmotionJSX } from '@emotion/react/types/jsx-namespace';
import { useFormik } from "formik";
import * as Styled from './Weather.styled'
import { validate } from './utils/validateSchema';


const Weather: React.FC = (): EmotionJSX.Element => {

  const [url, setUrl] = React.useState<string>('');
  const [press, setPress] = React.useState<boolean>(false);
  const [town, setTown] = React.useState<string>('')

  const formik = useFormik({
    initialValues: {
      searchField: '',
    },
    validate,
    onSubmit: (values, { resetForm }) => {
      setUrl(url => url =
        `http://api.weatherapi.com/v1/forecast.json?key=3e2b23202a804627a5c115738222701&q=${formik.values.searchField.trim()}&days=3&aqi=no&alerts=no`);
      setPress(true);
      setTown(values.searchField.slice(0, 1).toUpperCase() + values.searchField.slice(1).toLowerCase());
      resetForm();
    }
  })

  return (
    <Box sx={{
      display: 'flex',
      flexDirection: 'column',
      alignItems: 'center',
      width: "100%",
    }} >
      <Styled.Heading>Where do you want to check weather?</Styled.Heading>
      <form onSubmit={formik.handleSubmit}>
        <Box sx={{
          width: '40em',
          height: 150,
          mt: 3,
          display: 'flex',
          flexDirection: 'column',
          justifyContent: 'space-evenly'
        }}>
          <TextField
            placeholder='...maybe in Paris?'
            autoComplete='off'
            id='searchField'
            name='searchField'
            value={formik.values.searchField}
            onChange={formik.handleChange}
            error={(formik.touched.searchField && Boolean(formik.errors.searchField))}
            helperText={(formik.touched.searchField && formik.errors.searchField)}
          />
          <Button
            type='submit'
            sx={{
              width: 'fit-content',
              alignSelf: 'center',
            }}
            color='primary'
            variant="contained">
            Check weather
          </Button>
        </Box>
      </form>
      {press && <WeatherFetchData apiURL={url} townName={town} />}
      <span>
        {'Powered by '}
        <Link
          href="https://www.weatherapi.com/">WeatherAPI.com
        </Link>
      </span>
    </Box>
  );
}

export default Weather;
