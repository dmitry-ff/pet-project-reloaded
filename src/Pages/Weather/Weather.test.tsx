import { shallow } from "enzyme";
import Weather from "./Weather";
import renderer from 'react-test-renderer';

describe('Weather component', () => {

  it('should render one form element', () => {
    const component = shallow(<Weather />);
    const wrapper = component.find('form');
    expect(wrapper.length).toBe(1);
  });

})
export { }