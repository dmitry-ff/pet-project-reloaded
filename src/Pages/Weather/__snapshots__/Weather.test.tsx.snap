// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Weather component should render component with data array 1`] = `
<div
  className="MuiBox-root css-1mi3tt8"
>
  <h2
    className="css-us0qf1"
  >
    Where do you want to check weather?
  </h2>
  <form
    onSubmit={[Function]}
  >
    <div
      className="MuiBox-root css-fukaqj"
    >
      <div
        className="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <div
          className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoComplete="off"
            autoFocus={false}
            className="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            disabled={false}
            id="searchField"
            name="searchField"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="...maybe in Paris?"
            required={false}
            type="text"
            value=""
          />
          <fieldset
            aria-hidden={true}
            className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="css-nnbavb"
            >
              <span
                className="notranslate"
              >
                â€‹
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-lhxeoo-MuiButtonBase-root-MuiButton-root"
        disabled={false}
        onBlur={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="submit"
      >
        Check weather
      </button>
    </div>
  </form>
  <span>
    Powered by 
    <a
      className="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-1iwtne7-MuiTypography-root-MuiLink-root"
      href="https://www.weatherapi.com/"
      onBlur={[Function]}
      onFocus={[Function]}
    >
      WeatherAPI.com
    </a>
  </span>
</div>
`;
